Implement dd nearing completion

backwaterred opened the pull request:
Back in the day, I opened [#1861](https://github.com/uutils/coreutils/issues/1861) indicating that I would work on dd. I thought it would be a month or so to complete it, but I was off by _a small-ish margin_ ;-). Even though a few features are still missing, and some cleanup work remains, I am opening this pull request to bring dev work to the wider community, and basically stop hoarding the project.

Below is a list of todo items I am aware of.

- [x] Windows testing
- [x] MacOS testing
- [x] RedoxOS testing [edit]

Future work (to be created as separate issues)
- [ ] Additional testing. CLI testing for bad input. Trigger unit tests when tests performed from project root.
- [ ] Print stats in response to the INFO signal. I believe this is only available on BSD-style (i.e. MacOS included) systems. I'm not currently running one, and there doesn't appear to be a standard way to do this in Rust.
- [ ] A robust system-independent iflag=FLAG and oflag=FLAG implementation. Currently flags `[cio, nocache, nolinks, text, binary]` are not implemented at all, and `[direct, directory, dsync, sync, nonblock, noatime, noctty, nofollow]` are only expected to work on Linux. Testing is also a little light on this one.
- [ ] A good Ctrl-c handler. This should try to shut things down nicely at first, then get more aggressive after a second Ctrl-c.
- [ ] Deduplication of `dd_fileout(...)` and `dd_stdout(...)` (ideally there should just be one function). Fixing this will require adjusting the way that `fsync()` `fdatasync()` and `write_all()` are handled. The solution probably involves [dynamic dispatch](https://doc.rust-lang.org/book/ch17-02-trait-objects.html).
- [ ] Tune Clap to make better use of the lib's features. Ideas: default values, improved help text. Clarity on difference between traditional and modern arg passing (Clap should give us `--if foo.file` for free, and `if=foo.file` should continue to work), mention multiplier strings.
- [ ] Additional documentation for all the things.
- [ ] Address the newline question: Block & Unblock currently only accept LF as a newline character. Should this behaviour be made system dependent?

I'd like to continue working on this project, and am happy for any suggestions.

sylvestre commented:
Terrific, I guess you saw


```

---- test_dd::test_atime_updated stdout ----
current_directory_resolved: 
run: /target/x86_64-unknown-linux-gnu/debug/coreutils dd status=none if=this-file-exists-no-noatime.txt
thread 'test_dd::test_atime_updated' panicked at 'assertion failed: pre_atime != post_atime', /project/tests/by-util/test_dd.rs:301:5

---- test_dd::test_nocreat_causes_failure_when_outfile_not_present stdout ----
current_directory_resolved: 
run: /target/x86_64-unknown-linux-gnu/debug/coreutils dd conv=nocreat of=this-file-does-not-exist.txt
thread 'test_dd::test_nocreat_causes_failure_when_outfile_not_present' panicked at 'assertion failed: !fix.file_exists(&fname)', /project/tests/by-util/test_dd.rs:365:5


failures:
    test_dd::test_atime_updated
    test_dd::test_nocreat_causes_failure_when_outfile_not_present
```

sylvestre commented on src/uu/dd/src/datastructures.rs:
```diff
@@ -0,0 +1,163 @@
+use crate::conversion_tables::*;
```
```suggestion
// This file is part of the uutils coreutils package.
//
// For the full copyright and license information, please view the LICENSE
// file that was distributed with this source code.

// spell-checker:ignore (ToDO) ctable noerror

use crate::conversion_tables::*;
```

sylvestre commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1314 @@
+use std::thread;
+use std::time;
+
+const SYNTAX: &str = "dd [OPERAND]...\ndd OPTION";
```
```suggestion
```

sylvestre commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1314 @@
+    let mut flag = 0;
+
+    if oflags.direct {
+        flag |= libc::O_DIRECT;
```
please fix this error

sylvestre replied:
the CI was showing error (didn't exist)
you will be able to see them in `Files changed` once the CI is over

sylvestre commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1314 @@
+    dst: W,
+    obs: usize,
+    cflags: OConvFlags,
+    oflags: OFlags,
```
ditto

sylvestre commented on tests/by-util/test_dd.rs:
```diff
@@ -0,0 +1,702 @@
+use crate::common::util::*;
```
please run rustfmt on this file

tertsdiepraam reviewed (commented):
I posted a few comments about where things changed while this PR was in progress. Great work!

backwaterred commented:
Thanks for the feedback and kind words @tertsdiepraam and @sylvestre. I have incorporated much of your feedback.

Regarding ["please fix this error"](https://github.com/uutils/coreutils/pull/2474#discussion_r663533742) ["ditto"](https://github.com/uutils/coreutils/pull/2474#discussion_r663533766) by @sylvestre.  I'm not sure what you're suggesting here; can you please let me know what changes you had in mind?

Regarding [the failing tests](https://github.com/uutils/coreutils/pull/2474#issuecomment-873350705). This is a work in progress.

backwaterred commented:
Hello reviewers,

Thanks for your patience while I got all those rustfmt and clippy warnings sorted out. My local machine and the build machines were fighting. For a reason not clear to me, my local nightly toolchain and the one on the build machines produced different warnings (mine produced none) and formatting :-S

The only lint issue that remains is in sleep.rs (which is not in any way related to my branch). I'm happy to follow your suggestions on resolving (or not-resolving) this one.

At this point, I'd like to suggest that we merge this with master, and address the remaining items on the todo list [above](https://github.com/uutils/coreutils/pull/2474#issue-682888930) in separate issues.

tertsdiepraam commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+        .get_appropriate_unit(true)
+        .format(0);
+    let safe_millis = cmp::max(1, update.duration.as_millis());
+    let xfer_rate = Byte::from_bytes(1000 * (update.bytes_total / safe_millis))
```
I'm guessing this means `transfer_rate`?

backwaterred replied:
Sure is. I made this change.

tertsdiepraam commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+        let mut tmp;
+
+        for base in (1..buf.len()).step_by(2) {
+            tmp = buf[base];
+            buf[base] = buf[base - 1];
+            buf[base - 1] = tmp;
```
You could use [`swap`](https://doc.rust-lang.org/std/primitive.slice.html#method.swap)

backwaterred replied:
I made this change.

tertsdiepraam reviewed (changes requested):
Quite a few comments, but that's to be expected on such a large PR. I don't even think I understood the whole thing, so I'll have another look later. I'm quite unfamiliar with `dd`, so some of the "jargon" makes it hard to read. It would be great if you could improve documentation and variable names a bit to make it more readable. I do want to say that I think you did a phenomenal job. Everything looks very clean and nice! I also agree that we merge this before adding more features, because it's already very large :)

tertsdiepraam commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+            };
+            // Handle signals
+            #[allow(clippy::single_match)]
+            match sigval.load(Ordering::Relaxed) {
+                SIGUSR1_USIZE => {
+                    print_xfer_stats(&update);
+                }
+                // SIGINFO_USIZE => ...
+                _ => { /* no signals recv'd */ }
+            };
```
Instead of a match with one arm you could use `if let`:
```suggestion
            if let SIGNUSR1_USIZE = sigval.load(Ordering::Relaxed) {
                print_xfer_stats(&update);
            };
```
Although I guess you're keeping it as a `match` to allow for more signals later?

backwaterred replied:
That was the idea behind the match case yeah, but I think you bring up a good point. I removed the roughed out implementation for SIGINFO (the match case and a couple other lines). I think any future implementer will be just as likely to be confused by it than helped by it. A good suggestion.

tertsdiepraam commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+                opts.custom_flags(libc_flags);
+            }
+
+            let dst = opts.open(path)?;
+            Ok(dst)
```
```suggestion
            opts.open(path)
```

backwaterred replied:
This was actually due to a compiler message that went away when I did it this way. Addressed the message, and made this change.

tertsdiepraam commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+
+        Ok(buf)
+    } else {
+        // The following error should not happen, as it results from
+        // insufficient command line data. This case should be caught
+        // by the parser before making it this far.
+        // Producing this error is an alternative to risking an unwrap call
+        // on 'cbs' if the required data is not provided.
```
Great explanation!

backwaterred replied:
Thanks :-)

tertsdiepraam commented on src/uu/dd/src/datastructures.rs:
```diff
@@ -0,0 +1,197 @@
+
+/// Stores all Flags that apply to the input
+pub struct IFlags {
+    #[allow(dead_code)]
+    pub cio: bool,
+    #[allow(dead_code)]
+    pub direct: bool,
+    #[allow(dead_code)]
+    pub directory: bool,
+    #[allow(dead_code)]
+    pub dsync: bool,
+    #[allow(dead_code)]
+    pub sync: bool,
+    #[allow(dead_code)]
+    pub nocache: bool,
+    #[allow(dead_code)]
+    pub nonblock: bool,
+    #[allow(dead_code)]
+    pub noatime: bool,
+    #[allow(dead_code)]
+    pub noctty: bool,
+    #[allow(dead_code)]
+    pub nofollow: bool,
+    #[allow(dead_code)]
+    pub nolinks: bool,
+    #[allow(dead_code)]
+    pub binary: bool,
+    #[allow(dead_code)]
+    pub text: bool,
```
Are these still dead code?

tertsdiepraam replied:
Or are they linux only? In that case they should have an attribute that reflects that (i.e. `#[cfg(target_os = "linux")]`)

tertsdiepraam commented on src/uu/dd/src/datastructures.rs:
```diff
@@ -0,0 +1,197 @@
+/// Stores all Flags that apply to the output
+pub struct OFlags {
+    pub append: bool,
+    #[allow(dead_code)]
+    pub cio: bool,
+    #[allow(dead_code)]
+    pub direct: bool,
+    #[allow(dead_code)]
+    pub directory: bool,
+    #[allow(dead_code)]
+    pub dsync: bool,
+    #[allow(dead_code)]
+    pub sync: bool,
+    #[allow(dead_code)]
+    pub nocache: bool,
+    #[allow(dead_code)]
+    pub nonblock: bool,
+    #[allow(dead_code)]
+    pub noatime: bool,
+    #[allow(dead_code)]
+    pub noctty: bool,
+    #[allow(dead_code)]
+    pub nofollow: bool,
+    #[allow(dead_code)]
+    pub nolinks: bool,
+    #[allow(dead_code)]
+    pub binary: bool,
+    #[allow(dead_code)]
+    pub text: bool,
```
Are these still dead code?

tertsdiepraam commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+        let oflags = parseargs::parse_oflags(matches)?;
+        let seek = parseargs::parse_seek_amt(&obs, &oflags, matches)?;
+
+        if let Some(fname) = matches.value_of("of") {
```
Please use `options::OUTFILE`

tertsdiepraam commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+        let skip = parseargs::parse_skip_amt(&ibs, &iflags, matches)?;
+        let count = parseargs::parse_count(&iflags, matches)?;
+
+        if let Some(fname) = matches.value_of("if") {
```
Please use `options::INFILE`

tertsdiepraam commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+/// as specified by conv=unblock and cbs=N
+fn unblock(buf: Vec<u8>, cbs: usize) -> Vec<u8> {
+    // Local Helper Fns ----------------------------------------------------
+    #[inline]
+    fn build_blocks(buf: Vec<u8>, cbs: usize) -> Vec<Vec<u8>> {
+        let mut blocks = Vec::new();
+        let mut curr = buf;
+        let mut next;
+        let mut width;
+
+        while !curr.is_empty() {
+            width = cmp::min(cbs, curr.len());
+            next = curr.split_off(width);
+
+            blocks.push(curr);
+
+            curr = next;
+        }
+
+        blocks
+    }
```
If I'm reading this right this is equivalent to `buf.as_chunks(cbs).collect()`

tertsdiepraam commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+        blocks
+    }
+    // ---------------------------------------------------------------------
+    build_blocks(buf, cbs)
+        .into_iter()
+        .fold(Vec::new(), |mut unblocks, mut block| {
+            let block = if let Some(last_char_idx) = block.iter().rposition(|&e| e != b' ') {
+                block.truncate(last_char_idx + 1);
+                block.push(b'\n');
+
+                block
+            } else if let Some(b' ') = block.get(0) {
+                vec![b'\n']
+            } else {
+                block
+            };
+
+            unblocks.push(block);
+
+            unblocks
+        })
+        .into_iter()
+        .flatten()
+        .collect()
```
```suggestion
    buf.as_chuncks(cbs)
        .map(|mut block| {
            if let Some(last_char_idx) = block.iter().rposition(|&e| e != b' ') {
                block.truncate(last_char_idx + 1);
                block.push(b'\n');

                block
            } else if let Some(b' ') = block.get(0) {
                vec![b'\n']
            } else {
                block
            }
        })
        .flatten()
        .collect()
```

backwaterred replied:
Good spot on this one! I ended up removing a large amount of cruft by incorporating your feedback on this and [the previous](https://github.com/uutils/coreutils/pull/2474#discussion_r668709056) suggestion.

tertsdiepraam reviewed (commented):
Round 2 of comments :)

backwaterred commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+}
+
+#[cfg(target_os = "linux")]
+fn make_linux_iflags(oflags: &IFlags) -> Option<libc::c_int> {
```
Good catch. I made this change.

backwaterred commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+    }
+    // Local Helper Fns ----------------------------------------------------
+    #[inline]
+    fn apply_ct(buf: &mut [u8], ct: &ConversionTable) {
```
Agreed. I made this change.

backwaterred commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+/// Perform the copy/convert operations. File backed output version
+// Note: Some of dd's functionality depends on whether the output is actually a file. This breaks the Output<Write> abstraction,
+// and should be fixed in the future.
+fn dd_fileout<R: Read>(mut i: Input<R>, mut o: Output<File>) -> Result<(), Box<dyn Error>> {
```
I agree this is confusing, and I would love to fix it, but I don't currently know how. I ran into this problem early on in implementation, tried to fix it several times, but never quite got there. I've been doing some research over the past few days, and I think I can use OO features and dynamic dispatch to do what I want. I'm not really an OO programmer, so I skipped that chapter in the Rust book. I've added a comment for now, but this will probably be the fist issue I tackle when the code is merged. It's a major maintenance headache to keep dd_fileout the same as dd_stdout.

miDeb replied:
We can use traits to solve this. I don't know what the best approach would be, but here's one possiblity:
<details>
<summary> Patch </summary>

```
diff --git a/src/uu/dd/src/dd.rs b/src/uu/dd/src/dd.rs
index 2b44a47d..5f7af9ea 100644
--- a/src/uu/dd/src/dd.rs
+++ b/src/uu/dd/src/dd.rs
@@ -289,13 +289,19 @@ impl<R: Read> Input<R> {
     }
 }
 
+trait OutputTrait: Sized + Write {
+    fn new(matches: &Matches) -> Result<Self, Box<dyn Error>>;
+    fn fsync(&mut self) -> io::Result<()>;
+    fn fdatasync(&mut self) -> io::Result<()>;
+}
+
 struct Output<W: Write> {
     dst: W,
     obs: usize,
     cflags: OConvFlags,
 }
 
-impl Output<io::Stdout> {
+impl OutputTrait for Output<io::Stdout> {
     fn new(matches: &Matches) -> Result<Self, Box<dyn Error>> {
         let obs = parseargs::parse_obs(matches)?;
         let cflags = parseargs::parse_conv_flag_output(matches)?;
@@ -314,6 +320,111 @@ impl Output<io::Stdout> {
     }
 }
 
+impl<W: Write> Output<W>
+where
+    Self: OutputTrait,
+{
+    fn write_blocks(&mut self, buf: Vec<u8>) -> io::Result<WriteStat> {
+        let mut writes_complete = 0;
+        let mut writes_partial = 0;
+        let mut base_idx = 0;
+
+        while base_idx < buf.len() {
+            let next_blk = cmp::min(base_idx + self.obs, buf.len());
+            let plen = next_blk - base_idx;
+
+            match self.write(&buf[base_idx..next_blk])? {
+                wlen if wlen < plen => {
+                    writes_partial += 1;
+                    base_idx += wlen;
+                }
+                wlen => {
+                    writes_complete += 1;
+                    base_idx += wlen;
+                }
+            }
+        }
+
+        Ok(WriteStat {
+            writes_complete,
+            writes_partial,
+            bytes_total: base_idx.try_into().unwrap_or(0u128),
+        })
+    }
+
+    fn dd_out<R: Read>(mut self, mut i: Input<R>) -> Result<(), Box<dyn Error>> {
+        let mut rstat = ReadStat {
+            reads_complete: 0,
+            reads_partial: 0,
+            records_truncated: 0,
+        };
+        let mut wstat = WriteStat {
+            writes_complete: 0,
+            writes_partial: 0,
+            bytes_total: 0,
+        };
+        let start = time::Instant::now();
+        let bsize = calc_bsize(i.ibs, self.obs);
+
+        let prog_tx = {
+            let (tx, rx) = mpsc::channel();
+            thread::spawn(gen_prog_updater(rx, i.xfer_stats));
+            tx
+        };
+
+        while below_count_limit(&i.count, &rstat, &wstat) {
+            // Read/Write
+            let loop_bsize = calc_loop_bsize(&i.count, &rstat, &wstat, i.ibs, bsize);
+            match read_helper(&mut i, loop_bsize)? {
+                (
+                    ReadStat {
+                        reads_complete: 0,
+                        reads_partial: 0,
+                        ..
+                    },
+                    _,
+                ) => break,
+                (rstat_update, buf) => {
+                    let wstat_update = self.write_blocks(buf)?;
+
+                    rstat += rstat_update;
+                    wstat += wstat_update;
+                }
+            };
+            // Update Prog
+            prog_tx.send(ProgUpdate {
+                reads_complete: rstat.reads_complete,
+                reads_partial: rstat.reads_partial,
+                writes_complete: wstat.writes_complete,
+                writes_partial: wstat.writes_partial,
+                bytes_total: wstat.bytes_total,
+                records_truncated: rstat.records_truncated,
+                duration: start.elapsed(),
+            })?;
+        }
+
+        if self.cflags.fsync {
+            self.fsync()?;
+        } else if self.cflags.fdatasync {
+            self.fdatasync()?;
+        }
+
+        match i.xfer_stats {
+            Some(StatusLevel::Noxfer) | Some(StatusLevel::None) => {}
+            _ => print_xfer_stats(&ProgUpdate {
+                reads_complete: rstat.reads_complete,
+                reads_partial: rstat.reads_partial,
+                writes_complete: wstat.writes_complete,
+                writes_partial: wstat.writes_partial,
+                bytes_total: wstat.bytes_total,
+                records_truncated: rstat.records_truncated,
+                duration: start.elapsed(),
+            }),
+        }
+        Ok(())
+    }
+}
+
 #[cfg(target_os = "linux")]
 fn make_linux_oflags(oflags: &OFlags) -> Option<libc::c_int> {
     let mut flag = 0;
@@ -354,7 +465,7 @@ fn make_linux_oflags(oflags: &OFlags) -> Option<libc::c_int> {
     }
 }
 
-impl Output<File> {
+impl OutputTrait for Output<File> {
     fn new(matches: &Matches) -> Result<Self, Box<dyn Error>> {
         fn open_dst(
             path: &Path,
@@ -450,66 +561,6 @@ impl Write for Output<io::Stdout> {
     }
 }
 
-impl Output<io::Stdout> {
-    fn write_blocks(&mut self, buf: Vec<u8>) -> io::Result<WriteStat> {
-        let mut writes_complete = 0;
-        let mut writes_partial = 0;
-        let mut base_idx = 0;
-
-        while base_idx < buf.len() {
-            let next_blk = cmp::min(base_idx + self.obs, buf.len());
-            let plen = next_blk - base_idx;
-
-            match self.write(&buf[base_idx..next_blk])? {
-                wlen if wlen < plen => {
-                    writes_partial += 1;
-                    base_idx += wlen;
-                }
-                wlen => {
-                    writes_complete += 1;
-                    base_idx += wlen;
-                }
-            }
-        }
-
-        Ok(WriteStat {
-            writes_complete,
-            writes_partial,
-            bytes_total: base_idx.try_into().unwrap_or(0u128),
-        })
-    }
-}
-
-impl Output<File> {
-    fn write_blocks(&mut self, buf: Vec<u8>) -> io::Result<WriteStat> {
-        let mut writes_complete = 0;
-        let mut writes_partial = 0;
-        let mut base_idx = 0;
-
-        while base_idx < buf.len() {
-            let next_blk = cmp::min(base_idx + self.obs, buf.len());
-            let plen = next_blk - base_idx;
-
-            match self.write(&buf[base_idx..next_blk])? {
-                wlen if wlen < plen => {
-                    writes_partial += 1;
-                    base_idx += wlen;
-                }
-                wlen => {
-                    writes_complete += 1;
-                    base_idx += wlen;
-                }
-            }
-        }
-
-        Ok(WriteStat {
-            writes_complete,
-            writes_partial,
-            bytes_total: base_idx.try_into().unwrap_or(0u128),
-        })
-    }
-}
-
 /// Splits the content of buf into cbs-length blocks
 /// Appends padding as specified by conv=block and cbs=N
 fn block(buf: Vec<u8>, cbs: usize, rstat: &mut ReadStat) -> Vec<Vec<u8>> {
@@ -890,156 +941,6 @@ fn below_count_limit(count: &Option<CountType>, rstat: &ReadStat, wstat: &WriteS
     }
 }
 
-/// Perform the copy/convert operations. Stdout version
-// Note: Some of dd's functionality depends on whether the output is actually a file. This breaks the Output<Write> abstraction,
-// and should be fixed in the future.
-fn dd_stdout<R: Read>(mut i: Input<R>, mut o: Output<io::Stdout>) -> Result<(), Box<dyn Error>> {
-    let mut rstat = ReadStat {
-        reads_complete: 0,
-        reads_partial: 0,
-        records_truncated: 0,
-    };
-    let mut wstat = WriteStat {
-        writes_complete: 0,
-        writes_partial: 0,
-        bytes_total: 0,
-    };
-    let start = time::Instant::now();
-    let bsize = calc_bsize(i.ibs, o.obs);
-
-    let prog_tx = {
-        let (tx, rx) = mpsc::channel();
-        thread::spawn(gen_prog_updater(rx, i.xfer_stats));
-        tx
-    };
-
-    while below_count_limit(&i.count, &rstat, &wstat) {
-        // Read/Write
-        let loop_bsize = calc_loop_bsize(&i.count, &rstat, &wstat, i.ibs, bsize);
-        match read_helper(&mut i, loop_bsize)? {
-            (
-                ReadStat {
-                    reads_complete: 0,
-                    reads_partial: 0,
-                    ..
-                },
-                _,
-            ) => break,
-            (rstat_update, buf) => {
-                let wstat_update = o.write_blocks(buf)?;
-
-                rstat += rstat_update;
-                wstat += wstat_update;
-            }
-        };
-        // Update Prog
-        prog_tx.send(ProgUpdate {
-            reads_complete: rstat.reads_complete,
-            reads_partial: rstat.reads_partial,
-            writes_complete: wstat.writes_complete,
-            writes_partial: wstat.writes_partial,
-            bytes_total: wstat.bytes_total,
-            records_truncated: rstat.records_truncated,
-            duration: start.elapsed(),
-        })?;
-    }
-
-    if o.cflags.fsync {
-        o.fsync()?;
-    } else if o.cflags.fdatasync {
-        o.fdatasync()?;
-    }
-
-    match i.xfer_stats {
-        Some(StatusLevel::Noxfer) | Some(StatusLevel::None) => {}
-        _ => print_xfer_stats(&ProgUpdate {
-            reads_complete: rstat.reads_complete,
-            reads_partial: rstat.reads_partial,
-            writes_complete: wstat.writes_complete,
-            writes_partial: wstat.writes_partial,
-            bytes_total: wstat.bytes_total,
-            records_truncated: rstat.records_truncated,
-            duration: start.elapsed(),
-        }),
-    }
-    Ok(())
-}
-
-/// Perform the copy/convert operations. File backed output version
-// Note: Some of dd's functionality depends on whether the output is actually a file. This breaks the Output<Write> abstraction,
-// and should be fixed in the future.
-fn dd_fileout<R: Read>(mut i: Input<R>, mut o: Output<File>) -> Result<(), Box<dyn Error>> {
-    let mut rstat = ReadStat {
-        reads_complete: 0,
-        reads_partial: 0,
-        records_truncated: 0,
-    };
-    let mut wstat = WriteStat {
-        writes_complete: 0,
-        writes_partial: 0,
-        bytes_total: 0,
-    };
-    let start = time::Instant::now();
-    let bsize = calc_bsize(i.ibs, o.obs);
-
-    let prog_tx = {
-        let (tx, rx) = mpsc::channel();
-        thread::spawn(gen_prog_updater(rx, i.xfer_stats));
-        tx
-    };
-
-    while below_count_limit(&i.count, &rstat, &wstat) {
-        // Read/Write
-        let loop_bsize = calc_loop_bsize(&i.count, &rstat, &wstat, i.ibs, bsize);
-        match read_helper(&mut i, loop_bsize)? {
-            (
-                ReadStat {
-                    reads_complete: 0,
-                    reads_partial: 0,
-                    ..
-                },
-                _,
-            ) => break,
-            (rstat_update, buf) => {
-                let wstat_update = o.write_blocks(buf)?;
-
-                rstat += rstat_update;
-                wstat += wstat_update;
-            }
-        };
-        // Update Prog
-        prog_tx.send(ProgUpdate {
-            reads_complete: rstat.reads_complete,
-            reads_partial: rstat.reads_partial,
-            writes_complete: wstat.writes_complete,
-            writes_partial: wstat.writes_partial,
-            bytes_total: wstat.bytes_total,
-            records_truncated: rstat.records_truncated,
-            duration: start.elapsed(),
-        })?;
-    }
-
-    if o.cflags.fsync {
-        o.fsync()?;
-    } else if o.cflags.fdatasync {
-        o.fdatasync()?;
-    }
-
-    match i.xfer_stats {
-        Some(StatusLevel::Noxfer) | Some(StatusLevel::None) => {}
-        _ => print_xfer_stats(&ProgUpdate {
-            reads_complete: rstat.reads_complete,
-            reads_partial: rstat.reads_partial,
-            writes_complete: wstat.writes_complete,
-            writes_partial: wstat.writes_partial,
-            bytes_total: wstat.bytes_total,
-            records_truncated: rstat.records_truncated,
-            duration: start.elapsed(),
-        }),
-    }
-    Ok(())
-}
-
 // The compiler does not like Clippy's suggestion to use &str in place of &String here.
 #[allow(clippy::ptr_arg)]
 fn append_dashes_if_not_present(mut acc: Vec<String>, s: &String) -> Vec<String> {
@@ -1091,7 +992,7 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
             let (i, o) =
                 unpack_or_rtn!(Input::<File>::new(&matches), Output::<File>::new(&matches));
 
-            dd_fileout(i, o)
+            o.dd_out(i)
         }
         (false, true) => {
             let (i, o) = unpack_or_rtn!(
@@ -1099,7 +1000,7 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
                 Output::<File>::new(&matches)
             );
 
-            dd_fileout(i, o)
+            o.dd_out(i)
         }
         (true, false) => {
             let (i, o) = unpack_or_rtn!(
@@ -1107,7 +1008,7 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
                 Output::<io::Stdout>::new(&matches)
             );
 
-            dd_stdout(i, o)
+            o.dd_out(i)
         }
         (false, false) => {
             let (i, o) = unpack_or_rtn!(
@@ -1115,7 +1016,7 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
                 Output::<io::Stdout>::new(&matches)
             );
 
-            dd_stdout(i, o)
+            o.dd_out(i)
         }
     };
     match result {
diff --git a/src/uu/dd/src/dd_unit_tests/conversion_tests.rs b/src/uu/dd/src/dd_unit_tests/conversion_tests.rs
index 738ead5f..fdaadb48 100644
--- a/src/uu/dd/src/dd_unit_tests/conversion_tests.rs
+++ b/src/uu/dd/src/dd_unit_tests/conversion_tests.rs
@@ -153,7 +153,7 @@ fn all_valid_ascii_ebcdic_ascii_roundtrip_conv_test() {
         cflags: DEFAULT_CFO,
     };
 
-    dd_fileout(i, o).unwrap();
+    o.dd_out(i).unwrap();
 
     // EBCDIC->ASCII
     let test_name = "all-valid-ebcdic-to-ascii";
@@ -175,7 +175,7 @@ fn all_valid_ascii_ebcdic_ascii_roundtrip_conv_test() {
         cflags: DEFAULT_CFO,
     };
 
-    dd_fileout(i, o).unwrap();
+    o.dd_out(i).unwrap();
 
     // Final Comparison
     let res = File::open(&tmp_fname_ea).unwrap();
diff --git a/src/uu/dd/src/dd_unit_tests/mod.rs b/src/uu/dd/src/dd_unit_tests/mod.rs
index 49edadd9..09ec5222 100644
--- a/src/uu/dd/src/dd_unit_tests/mod.rs
+++ b/src/uu/dd/src/dd_unit_tests/mod.rs
@@ -121,7 +121,7 @@ macro_rules! make_spec_test (
         #[test]
         fn $test_id()
         {
-            dd_fileout($i,$o).unwrap();
+            $o.dd_out($i).unwrap();
 
             let res = File::open($tmp_fname).unwrap();
             // Check test file isn't empty (unless spec file is too)
```
</details>

backwaterred replied:
From looking at the changes, this looks really promising. How do you (@miDeb) want to proceed with the change? I'm happy to get this PR merged then let you tackle this issue. Or is there a way for you to merge with this PR while it's in progress?

miDeb replied:
I think it's ok to merge this with the code duplication, since it's not a bug but a code-quality thing. I would open a follow-up PR afterwards with this change.

backwaterred commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+        let mut writes_partial = 0;
+        let mut base_idx = 0;
+
+        while base_idx < buf.len() {
+            let next_blk = cmp::min(base_idx + self.obs, buf.len());
+            let plen = next_blk - base_idx;
+
+            match self.write(&buf[base_idx..next_blk])? {
+                wlen if wlen < plen => {
+                    writes_partial += 1;
+                    base_idx += wlen;
+                }
+                wlen => {
+                    writes_complete += 1;
+                    base_idx += wlen;
+                }
+            }
+        }
```
Agreed on both points. These changes are completed.

backwaterred commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+        //     debug_println!("Internal dd Warning: Unable to register SIGINFO handler \n\t{}", e);
+        // }
+        if !posixly_correct() {
+            if let Err(e) =
```
I'm not sure about this one. Changing the variable name, (and presumably also removing the print statement below it) means a silent error if the signal handler fails to register.

Or maybe you meant to change a different line (I think 794?).

tertsdiepraam replied:
Oh maybe, I meant the one that caused the unused variable warning

miDeb replied:
The warning is because the print is only there in debug mode, so in release mode `e` is unused. Maybe we want to print the warning in release mode too? Otherwise I think it's fine to ignore this (add `#[allow(unused_variables)]`)

backwaterred replied:
I think miDeb is right about the issue.

I've adjusted this to print the warning on all builds unless `status=none` is specified.

miDeb commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+    }
+    // --------------------------------------------------------------
+    move || {
+        const SIGUSR1_USIZE: usize = signal::SIGUSR1 as usize;
```
It looks like `SIGUSR1` is not available on windows.
Where are we receiving `SIGUSR1` signals from?

backwaterred replied:
Good call. I've added a guard to the relevant code so it's only present on Linux.

miDeb commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+    Ok(())
+}
+
+// The compiler does not like Clippy's suggestion to use &str in place of &String here.
+#[allow(clippy::ptr_arg)]
+fn append_dashes_if_not_present(mut acc: Vec<String>, s: &String) -> Vec<String> {
+    if Some("--") != s.get(0..=1) {
+        acc.push(format!("--{}", s));
+    }
+    acc
+}
```
```suggestion
fn append_dashes_if_not_present(mut acc: Vec<String>, mut s: String) -> Vec<String> {
    if !s.starts_with("--") {
        s.insert_str(0, "--");
        acc.push(s);
    }
    acc
}
```
And change the `iter()` in uumain to `into_iter()`. I don't fully understand the context here, but do we want to ignore args that start with `--` already?

backwaterred replied:
I like it :-) (I made this change).

miDeb commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+    };
+    match result {
+        Ok(_) => RTN_SUCCESS,
+        Err(e) => {
```
Same as with the other `debug_println!`, this warning is because `e` is unused in release mode.

miDeb commented on src/uu/dd/src/dd_unit_tests/block_unblock_tests.rs:
```diff
@@ -0,0 +1,357 @@
+        res,
+        vec![
+            vec![0u8, 1u8, 2u8, 3u8],
+            // vec![4u8, SPACE, SPACE, SPACE],
```
Is this comment still needed?

backwaterred replied:
A reasonable question :-P. I added some additional comments to the test file to explain the commented lines. They're just present for reference, since these parts of the test case should be truncated by the block function.

miDeb commented on src/uu/dd/src/dd_unit_tests/block_unblock_tests.rs:
```diff
@@ -0,0 +1,357 @@
+
+use super::*;
+
+const NL: u8 = '\n' as u8;
+const SPACE: u8 = ' ' as u8;
```
```suggestion
const NL: u8 = b'\n';
const SPACE: u8 = b' ';
```
Nit: you can make ASCII characters bytes with `b`.

backwaterred replied:
:+1:

miDeb commented on src/uu/dd/src/dd_unit_tests/mod.rs:
```diff
@@ -0,0 +1,143 @@
+    skip_bytes: false,
+};
+
+// const DEFAULT_OFLAGS: OFlags = OFlags {
```
Do we still need this? An alternative to adding these default consts would be to add `#[derive(Default)]` to `OConvFlags`, `IFlags` and `OFlags` directly, which will allow to use e.g. `OConvFlags::default()` which also fills all fields with `false`.

backwaterred replied:
Good suggestion. I've implemented this.

miDeb commented on src/uu/dd/src/parseargs.rs:
```diff
@@ -0,0 +1,761 @@
+    CountBytes,
+    SkipBytes,
+    // Either
+    #[allow(unused)]
+    Cio,
+    #[allow(unused)]
+    Direct,
+    #[allow(unused)]
+    Directory,
+    #[allow(unused)]
+    Dsync,
+    #[allow(unused)]
+    Sync,
+    #[allow(unused)]
+    NoCache,
+    #[allow(unused)]
+    NonBlock,
+    #[allow(unused)]
+    NoATime,
+    #[allow(unused)]
+    NoCtty,
+    #[allow(unused)]
+    NoFollow,
+    #[allow(unused)]
+    NoLinks,
+    #[allow(unused)]
+    Binary,
+    #[allow(unused)]
```
Are these still unused?

backwaterred replied:
Removed for the most recent build. Looks like no errors were triggered.

miDeb commented on src/uu/dd/src/parseargs.rs:
```diff
@@ -0,0 +1,761 @@
+
+/// Parse IFlags struct from CL-input
+pub fn parse_iflags(matches: &Matches) -> Result<IFlags, ParseError> {
+    let mut cio = false;
```
I think you can create a `IFlags` struct at the beginning of the function and modify its fields, instead of using these temprorary variables. As I mentioned previously you can derive default for `IFlags` to enable a `IFlags::default()` constructor.

backwaterred replied:
I'm hesitant to apply this one. My feelings is that the Default trait is great for tests and other situations when the value is relatively unimportant. However, the parser, however, depends on the flags being set to false when they're not explicitly provided, so

miDeb replied:
Default will not change the default value for bool. This would be a breaking change for rust, so I'm really sure it will never happen. Note that you can also manually implement default (`impl Default for IFlags `) if you need different defaults. However, since all defaults are `false`, deriving default would IMHO be the best solution.

Feel free to keep it this way, the current version is definitely good too.

backwaterred replied:
The implementation with Default looks much cleaner. You bring up a good point that I could simply implement Default if I wanted to continue my usual controlled lifestyle. However, I'm going to live on the edge and `#[derive(Default)`. Thanks for the suggestion.

miDeb commented on src/uu/dd/src/parseargs/unit_tests.rs:
```diff
@@ -0,0 +1,401 @@
+test_byte_parser!(test_bytes_G, "3G", 3 * 1024 * 1024 * 1024);
+test_byte_parser!(test_bytes_Gi, "3GiB", 3 * 1024 * 1024 * 1024);
+
+test_byte_parser!(test_bytes_TB, "4TB", 4 * 1000 * 1000 * 1000 * 1000);
```
The errors about the overflow is because on 32 bit systems this will overflow an usize. To fix you can add `    #[cfg(target_pointer_width = "64")]` on the offending lines, which will only enable the test on 64 bit.

backwaterred replied:
Agreed, however testsdiepraam mentioned that `uucore::parse_size` is an option. I've switched to that instead.

backwaterred replied:
Actually, I mis-read this. I have now made this change.

backwaterred commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+            }
+        };
+        // Update Prog
+        prog_tx.send(ProgUpdate {
+            reads_complete: rstat.reads_complete,
+            reads_partial: rstat.reads_partial,
+            writes_complete: wstat.writes_complete,
+            writes_partial: wstat.writes_partial,
+            bytes_total: wstat.bytes_total,
+            records_truncated: rstat.records_truncated,
+            duration: start.elapsed(),
+        })?;
```
Yes it can, and now it is :-)

backwaterred commented on src/uu/dd/src/dd_unit_tests/block_unblock_tests.rs:
```diff
@@ -0,0 +1,357 @@
+macro_rules! rs (
+    () =>
+    {
+        ReadStat {
+            reads_complete: 0,
+            reads_partial: 0,
+            records_truncated: 0,
+        }
```
Oh my, Rust has features beyond my wildest imagination! Thanks for pointing this out.

backwaterred commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+    bsize: usize,
+) -> Result<(ReadStat, Vec<u8>), Box<dyn Error>> {
+    // Local Predicate Fns -----------------------------------------------
+    #[inline]
+    fn is_conv<R: Read>(i: &Input<R>) -> bool {
+        i.cflags.ctable.is_some()
+    }
```
I read the article the same way. I've removed the redundant `#[inline]` tags.

backwaterred commented on src/uu/dd/src/parseargs.rs:
```diff
@@ -0,0 +1,761 @@
+    }
+}
+
+fn parse_bytes_with_opt_multiplier(s: &str) -> Result<usize, ParseError> {
+    match s.find(char::is_alphabetic) {
+        Some(idx) => {
+            let base = parse_bytes_only(&s[..idx])?;
+            let mult = parse_multiplier(&s[idx..])?;
+
+            if let Some(bytes) = base.checked_mul(mult) {
+                Ok(bytes)
+            } else {
+                Err(ParseError::MultiplierStringWouldOverflow(s.to_string()))
+            }
+        }
+        _ => parse_bytes_only(s),
+    }
+}
```
I made this change.

backwaterred commented on src/uu/dd/src/parseargs/unit_tests.rs:
```diff
@@ -0,0 +1,401 @@
+#[cfg(not(target_os = "linux"))]
+#[test]
+fn unimplemented_flags_should_error_non_unix() {
+    let mut unfailed = Vec::new();
```
Ok :-P

backwaterred commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1296 @@
+// For the full copyright and license information, please view the LICENSE
+// file that was distributed with this source code.
+
+/* cspell:disable */
```
Okey, I'll give it a go.

miDeb commented on src/uu/dd/Cargo.toml:
```diff
@@ -0,0 +1,32 @@
+[dependencies]
+byte-unit = "4.0"
+clap = { version = "2.33", features = [ "wrap_help" ] }
+debug_print = "1.0"
```
You don't use this dependency anymore. Can you remove it?

backwaterred replied:
Nice catch. It is removed.

miDeb commented on src/uu/dd/src/datastructures.rs:
```diff
@@ -0,0 +1,171 @@
+/// Stores all Conv Flags that apply to the input
+#[derive(Debug, Default, PartialEq)]
+pub struct IConvFlags {
+    pub ctable: Option<&'static ConversionTable>,
```
(You can ignore the compile error in CI for this, we are about to increase our minimum rust version requirement)

backwaterred replied:
Thank the stars! This has been giving me grief for a while, but it's never been a big enough issue to actually spend time addressing it (I was always able to work around it). Seems that not dealing with the problem has finally paid off :-P

miDeb commented on src/uu/dd/src/datastructures.rs:
```diff
@@ -0,0 +1,171 @@
+    pub sync: bool,
+    pub nocache: bool,
+    pub nonblock: bool,
+    pub noatime: bool,
```
Please add the [unknown words](https://github.com/uutils/coreutils/runs/3131291167?check_suite_focus=true) to spell checker allow list, or choose different words that don't trigger the spell checker.

backwaterred replied:
I believe I got them all. We'll see next push :crossed_fingers:

miDeb commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1191 @@
+use byte_unit::Byte;
+use clap::{self, crate_version};
+use gcd::Gcd;
+use signal_hook::consts::signal;
```
Since this import is unused on platforms other than linux, to get rid of the warnings you have to add `#[cfg(target_os = "linux")]` on it too. This also applies to the other "unused" warnings in this file.

backwaterred replied:
Gotcha. This should be taken care of.

miDeb commented on src/uu/dd/cspell.json:
```diff
@@ -0,0 +1,67 @@
+{
```
It looks like CI is not picking this file up. We usually put the `words` as ignores on each file separately. Alternatively, you could add all of these words to `.vscode/cspell.dictionaries/jargon.wordlist.txt`. You can add the `ignorePaths` to the global `.vscode/cspell.json` file

backwaterred replied:
Ha! I didn't think to look in the .vscode folder for the spell checker overrides. I've added some dd jargon to the jargon file and an exception for the unit-tests. Thanks for pointing this out.

miDeb commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1185 @@
+    fn posixly_correct() -> bool {
+        env::var("POSIXLY_CORRECT").is_ok()
+    }
+    fn register_signal_handlers(sigval: Arc<AtomicUsize>) -> Result<(), Box<dyn Error>> {
```
So this warns about `sigval` being unused on win/macos. To avoid this, I'd mark the whole function definition `#[cfg(target_os = "linux")]`, as well as the function call.

backwaterred replied:
I'm hoping to have `register_signal_handlers(...)` perform the work to register handlers for all platforms. To allow this extension, I think `#[allow(unused_variables)]` would be a better choice. I've added a comment to this effect in the file. 

We can also go a different route with this if you prefer.

miDeb replied:
No, do as you prefer

miDeb commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1185 @@
+    Ok(())
+}
+
+fn append_dashes_if_not_present(mut acc: Vec<String>, mut s: String) -> Vec<String> {
+    if !s.starts_with("--") && !s.starts_with('-') {
+        s.insert_str(0, "--");
+    }
+    acc.push(s);
+    acc
+}
```
So if I'm understanding this correctly this is to work around the fact that `dd` does not take arguments in the `--option` form but as `option=` (without the dashes). However this leads to some problems: Imagine invoking `dd` like this: `dd if foo.file`. This will get translated to `dd --if --foo.file`. This will result in an error, but the error will say that `Found argument '--foo.file' which wasn't expected`. This is not the right error message.

miDeb commented on src/uu/dd/src/parseargs/unit_tests.rs:
```diff
@@ -0,0 +1,565 @@
+// TODO: This should work, but Clap doesn't seem to understand it. Leaving it for now since the traditional dd if=foo.file works just fine.
+fn test_all_top_level_args_leading_dashes_sep_by_spaces() {
+    let args = vec![
+        String::from("dd"),
+        String::from("--if foo.file"),
+        String::from("--of bar.file"),
+        String::from("--ibs 10"),
+        String::from("--obs 10"),
+        String::from("--cbs 1"),
+        String::from("--bs 100"),
+        String::from("--count 2"),
+        String::from("--skip 2"),
+        String::from("--seek 2"),
+        String::from("--status progress"),
+        String::from("--conv ascii,ucase"),
+        String::from("--iflag count_bytes,skip_bytes"),
+        String::from("--oflag append,seek_bytes"),
```
This fails because args are split at whitespace when typing them in a terminal. You'll have to do this here too, e.g. not `"--if foo.file"` but `"--if"` and then `"foo.file"` as another string.

However, this will also not pass the test because of the logic in `append_dashes_if_not_present`. For example, `--if foo.file --of bar.file` will be turned into `--if --foo.file --of --bar.file`.

backwaterred replied:
Glad for your help on this one. You've definitely helped to clarify some things I was struggling with. 

As a side note, I was hoping to be able to support both `dd --if foo.file ...` and `dd if=foo.file ...` since it felt like Clap was going to do that easily, but I didn't anticipate the issues you mentioned [above](https://github.com/uutils/coreutils/pull/2474#discussion_r675782227). 

To resolve this and the issue above, I've set all args that take values to require equals. Both `dd --if=foo.file ...` and `dd if=foo.file ...` appear to work and provide sensible error messages. However, there is probably more work to be done here in the future to make it a really excellent solution.

miDeb replied:
Yeah, it doesn't need to be perfect from the beginning. We can always improve it

miDeb commented on src/uu/dd/src/parseargs/unit_tests.rs:
```diff
@@ -0,0 +1,565 @@
+}
+
+#[test]
+fn parse_iflag_tokens() {
```
This test fails on win/macOS.

backwaterred replied:
Fixed it.

miDeb commented on src/uu/dd/src/parseargs/unit_tests.rs:
```diff
@@ -0,0 +1,565 @@
+
+#[test]
+fn parse_oflag_tokens() {
+    let exp = vec![
```
This one too.

backwaterred replied:
Fixed this one too.

miDeb reviewed (commented):
Some more comments :slightly_smiling_face:

backwaterred commented:
May I get the pending build actions approved?

sylvestre commented on src/uu/dd/src/dd.rs:
```diff
@@ -0,0 +1,1244 @@
+use std::path::Path;
+#[cfg(target_os = "linux")]
+use std::sync::atomic::Ordering;
+use std::sync::{atomic::AtomicUsize, mpsc, Arc};
```
```suggestion
#[cfg(target_os = "linux")]
use std::sync::{atomic::AtomicUsize, Arc};
use std::sync::mpsc, Arc;
```

backwaterred commented:
I've addressed the block/unblock tests that fail on Windows... by ignoring them. I was working on a solution, but it was just big enough to feel like it should be a separate issue, so I've reverted my changes.

Some questions came up as I was implementing this feature. If there are certain answers to them feel free to link or give them here. If there are no certain answers and a bit of discussion is required, then it would probably be best to have this discussion in the issue itself. Either way, I'll link this comment in the issue when it's created, and link this comment for reference. Questions are below:

- What set of newline markers should this feature (block & unblock) support?
    - Does the project already has a stance on newlines?
- Should support be determined by the platform uutils was compiled on (ie. recognize only CR+LF on Windows, only LF on Linux, LF and maybe? CR on MacOS) or take Python's stance of [universal newlines](https://docs.python.org/3/glossary.html#term-universal-newlines) (recognize all newlines on all platforms).

sylvestre commented:
Code coverage and clippy are failing on Windows.
Could you please just disable these macros for this platform and we will land it? :)
this PR is becoming quite big :)

backwaterred commented:
Agreed. Those issues should be addressed in the most recent push.

sylvestre commented:
Excellent! many thanks :)
